### PR TITLE
test: Validate behavior on cancel, amend, and backend methods for Product Bundles

### PR DESCRIPTION
**Pull request by Prateek8848**
🧪 Title:
Packed Item — Validate behavior on cancel, amend, and backend methods for Product Bundles

📝 Description:
Test Summary:
The Packed Item Doctype plays a critical role in managing bundled products (via Product Bundle) and their interaction with delivery workflows. These tests validate backend logic around Delivery Note creation, cancellation, amendment, and system-level triggers like on_doctype_update.

By simulating real-world transaction behavior — including cancellation and amendment of documents and method-level validation — the test ensures consistency and robustness in logistics and inventory flows.

📄 Doctypes Covered:
Packed Item

Delivery Note

Product Bundle

Item

Customer

Warehouse

✅ Test Cases Implemented:
🔸 TC_SCK_415: test_update_packed_item_from_cancelled_doc_TC_SCK_415
Creates a Delivery Note with a bundled item. Submits and cancels the document, then creates an amended version from the cancelled one.

Expected Results:

Initial Delivery Note submission succeeds.

Cancelling the note sets status to "Cancelled".

A new copy of the note (amendment) can be created successfully.

🔸 TC_SCK_416: test_on_doctype_update_TC_SCK_416
Tests on_doctype_update trigger on the Packed Item doctype, along with validation from get_items_from_product_bundle. Attempts to get packed item data without specifying company.

Expected Results:

on_doctype_update() executes without error.

Calling get_items_from_product_bundle() without company triggers a frappe.ValidationError.

Error message: "Please specify Company"

🛠️ Key Enhancements:
Validated cancellation and amendment logic for Delivery Notes tied to Product Bundles.

Reinforced backend validation to prevent missing context (e.g., company) in method-level calls.

Ensured clean data creation using make_test_item, create_customer, and test warehouse setup.

📌 Scenarios Covered:
Delivery Note submission and cancellation flow.

Amending a cancelled document for reuse.

Backend validation during packed item method execution.

System method behavior (on_doctype_update, get_items_from_product_bundle) under edge case inputs.

🧪 Technology Used:
Python unittest framework

Frappe test utilities:

frappe.get_doc, frappe.copy_doc, frappe.db.exists

create_customer, make_test_item

self.assertRaises(frappe.ValidationError)

🔗 Linked Feature/Bug:
N/A

🆔 Issue ID:
N/A

### CODE REVIEW

File: erpnext/stock/doctype/packed_item/test_packed_item.py [hunk 1]
```diff
@@ -53,6 +55,119 @@ def setUpClass(cls) -> None:
...
+		if not frappe.db.exists("Customer", "_Test Customer"):
+			create_customer("_Test Customer", currency="INR")
+		item = "test packed item"
+		item = make_test_item(item)
+		item.item_group = "Products"
+		item.is_stock_item = 0
+		item.is_fixed_asset = 0
+		item.auto_create_assets = 0
+		item.save()
+		assert frappe.db.exists("Item", "Test Item")
```
Comment (aasif-patel) [inline erpnext/stock/doctype/packed_item/test_packed_item.py]: creating "test packed item" but assert checking "Test Item"?

File: erpnext/stock/doctype/packed_item/test_packed_item.py [hunk 2]
```diff
@@ -53,6 +55,119 @@ def setUpClass(cls) -> None:
...
+		self.assertEqual(dn.status, "To Bill", f"Unexpected status after submit: {dn.status}")
+		dn.reload()
+		dn.cancel()
+		self.assertEqual(dn.docstatus, 2, "Delivery Note was not cancelled")
+		self.assertEqual(dn.status, "Cancelled", f"Expected status 'Cancelled', got {dn.status}")
+		amended_dn = frappe.copy_doc(dn)
+		amended_dn.amended_from = dn.name
+		amended_dn.docstatus = 0
+		amended_dn.name = None  # allow system to generate new name
+		amended_dn.insert()
```
Comment (aasif-patel) [inline erpnext/stock/doctype/packed_item/test_packed_item.py:105]: why are we amending the cancelled delivery note?

File: erpnext/stock/doctype/packed_item/test_packed_item.py [hunk 3]
```diff
@@ -53,6 +55,119 @@ def setUpClass(cls) -> None:
...
+		if not frappe.db.exists("Customer", "_Test Customer"):
+			create_customer("_Test Customer", currency="INR")
+		item = "test packed item"
+		item = make_test_item(item)
+		item.item_group = "Products"
+		item.is_stock_item = 0
+		item.is_fixed_asset = 0
+		item.auto_create_assets = 0
+		item.save()
+		assert frappe.db.exists("Item", "Test Item")
```
Comment (aasif-patel) [inline erpnext/stock/doctype/packed_item/test_packed_item.py]: same here creating and checking diff item

File: erpnext/stock/doctype/packed_item/test_packed_item.py [hunk 4]
```diff
@@ -53,6 +55,119 @@ def setUpClass(cls) -> None:
...
+						"stock_uom": "Nos",
+					}
+				],
+			}
+		).insert()
+		on_doctype_update()
+		dn.submit()
+		msg = "Please specify Company"
+		with self.assertRaises(frappe.ValidationError, msg=msg):
+			get_items_from_product_bundle(json.dumps(product_bundle.items[0].as_dict()))
```
Comment (aasif-patel) [inline erpnext/stock/doctype/packed_item/test_packed_item.py]: use assertIn for the validation assertRaises passing the test for all the validation,

**Review by aasif-patel — CHANGES_REQUESTED** [review]

**Review by aasif-patel — APPROVED** [review]

**Comment by swathibollineni2** [comment]
@Prateek8848  created new issue https://github.com/8848digital/erpnext/issues/2701
TC_SCK_415, TC_SCK_416 - ValidationError: Company GSTIN is a mandatory field for GST Transactions
